### PR TITLE
8242451: ensure semantics of non-capturing lambdas are preserved independent of execution mode

### PR DESCRIPTION
**Pull request by gilles-duboscq**
[JDK-8232806](https://bugs.openjdk.java.net/browse/JDK-8232806) introduced the jdk.internal.lambda.disableEagerInitialization system property to be able to disable eager initialization of lambda classes. This was necessary to prevent side effects of class initializers triggered by such initialization in the context of the GraalVM native image tool.

However, the change as it is implemented means that the behaviour of non-capturing lambdas depends on the value of `disableEagerInitialization`: when it is false (the default) such lambdas are actually a singleton while when it is true, a fresh instance is returned every time.

Programs should definitely _not_ rely on reference equality since the Java spec does not guarantee it. However, in order to separate concern and ease debugging such bad programs, `disableEagerInitialization` shouldn't influence the singleton vs. fresh instance behaviour of lambdas in either direction.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8242451](https://bugs.openjdk.java.net/browse/JDK-8242451): ensure semantics of non-capturing lambdas are preserved independent of execution mode

### Reviewers
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/93/head:pull/93`
`$ git checkout pull/93`

### CODE REVIEW

File: test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java
Comment (lahodaj) [inline test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java:29]: I would suggest to consider having the test under test/jdk/(java/lang/invoke/lambda), not under test/langtools/tools/javac.

File: src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java
Comment (mlchung) [inline src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java:215]: Nit: it'd be good to wrap this long line.  There are a couple long lines in this patch.

File: test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java
Comment (mlchung) [reply test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java:29]: It's a good suggestion as `disableEagerInitialization` support is not part of javac.

**Review by mlchung — APPROVED** [review]
Looks good.   I agree with Jan's suggestion that it's good to move the test to test/jdk/java/lang/invoke/lambda which is a better home for it.

File: test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java
Comment (gilles-duboscq) [reply test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java:29]: OK makes sense. I guess it's still good to clean the test comments of the old `disableEagerInitialization` references?

File: src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java
Comment (gilles-duboscq) [reply src/java.base/share/classes/java/lang/invoke/InnerClassLambdaMetafactory.java]: I have wrapped some lines that were longer than the typical line in this file.
Let me know if the wrapping looks good to you.

File: test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java
Comment (gilles-duboscq) [reply test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java]: I have created a new test under `test/jdk/java/lang/invoke/lambda` and cleaned up `LambdaTest6.java`

File: test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java
Comment (mlchung) [inline test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java:33]: This line was added by JDK-8232806 (https://hg.openjdk.java.net/jdk/jdk/rev/27c2d2a4b695).
I assume you want to move the test case for JDK-8232806 to test/jdk/java/lang/invoke?  If so,
BridgeMethod.java should be looked at too.

Comment (gilles-duboscq) [reply test/langtools/tools/javac/lambda/lambdaExpression/LambdaTest6.java:33]: I have removed all the `disableEagerInitialization` tests from `BridgeMethod.java`. It is now restored to its pre-JDK-8232806 state.

**Comment by mlchung** [comment]
Thanks for updating BridgeMethod.java.  

I expected that the new LambdaEagerInitTest.java will be updated to verify the capturing
 lambda case that does not have the static ``LAMBDA_INSTANCE$` field.  BTW,
this new regression test should be moved to `test/jdk/java/lang/invoke/lambda/` 
instead of `test/jdk/java/lang/invoke`.  The test uses `assert`.   Note that java 
assertion is not enabled by default.  So regression tests should do an explicit check
and throw runtime exception when the test fails.    You can also use the JDK test 
library `jdk.test.lib.Asserts` or make this a testng test to use TestNG Asserts API.

I updated and improved the test for your reference (send you separately).

**Review by mlchung — APPROVED** [review]
Looks good.  Thanks for the update.

**Comment by gilles-duboscq** [comment]
Thanks @mlchung. Do I need a second review?

**Comment by mlchung** [comment]
> Thanks @mlchung. Do I need a second review?

No need.  You can integrate once you run the regression tests (I usually run tier1-tier3).

**Pull request closed without merging**
